samples/kubernetes/graceful-shutdown/graceful-shutdown.md

```diff
@@ -53,50 +53,66 @@ app.Use((httpContext, next) =>
+The reason for this behavior is that ingress and Kubernetes control plane are separate entities, operating independently. When Kubernetes decides to shut down some pod(s), there is inevitable delay between control plane taking a pod out of Service backend, and ingress "noticing" that this has happened and actually stopping routing requests to that pod. So it is useful to introduce some delay between the time the pod receivers `SIGTERM` signal, and the time when it stops responding to requests. You can make it happen by replacing the default `IHostLifetime` implementation (which shuts down the application immediately upon receiving `SIGTERM` signal):
 
 ```csharp
-public class DelayedLinkedTokenSource
+using System.Runtime.InteropServices;
+
+public class DelayedShutdownHostLifetime : IHostLifetime, IDisposable
 {
-    private CancellationTokenSource _cts;
+    private IHostApplicationLifetime _applicationLifetime;
+    private TimeSpan _delay;
+    private IEnumerable<IDisposable> _disposables;
+
+    public DelayedShutdownHostLifetime(IHostApplicationLifetime applicationLifetime, TimeSpan delay) { 
+        _applicationLifetime = applicationLifetime;
+        _delay = delay;
+    }
 
-    public DelayedLinkedTokenSource(CancellationToken source, TimeSpan delay)
+    public Task StopAsync(CancellationToken cancellationToken)
     {
-        _cts = new CancellationTokenSource();
-        source.Register(() => _cts.CancelAfter(delay));
+        return Task.CompletedTask;
     }
 
-    public CancellationToken Token => _cts.Token;
-}
+    public Task WaitForStartAsync(CancellationToken cancellationToken)
+    {
+        _disposables = new IDisposable[]
+        {
+            PosixSignalRegistration.Create(PosixSignal.SIGINT, HandleSignal),
+            PosixSignalRegistration.Create(PosixSignal.SIGQUIT, HandleSignal),
+            PosixSignalRegistration.Create(PosixSignal.SIGTERM, HandleSignal)
+        };
+        return Task.CompletedTask;
+    }
 
-public class ApplicationStoppingTokenSource: DelayedLinkedTokenSource
-{
-    public static readonly TimeSpan ShutdownDelay = TimeSpan.FromSeconds(3); // Change the delay as appropriate for your service.
+    protected void HandleSignal(PosixSignalContext ctx)
+    {
+        ctx.Cancel = true;
+        Task.Delay(_delay).ContinueWith(t => { _applicationLifetime.StopApplication(); });
+    }
 
-    public ApplicationStoppingTokenSource(IHostApplicationLifetime hostLifetime) :
-        base(hostLifetime.ApplicationStopping, ShutdownDelay)
-    { }
+    public void Dispose()
+    {
+        if (_disposables != null)
+        {
+            foreach (var disposable in _disposables) { disposable.Dispose(); }
```

Comment:
We should follow our own coding guidelines and break this code into multiple lines.

Reply:
Updated